bootstore/src/db/mod.rs

```diff
@@ -128,16 +173,170 @@ impl Db {
+    // and so we only call this from the `initialize` method.
+    //
+    // Since we only allow a single row, we just delete the existing rows
+    // and insert the new one.
+    //
+    // Return the old rack UUID if one exists
+    //
+    // TODO: Add a DB trigger to ensure only 1 row can exist at a time?
+    fn initialize_rack_uuid(
+        &self,
+        tx: &mut SqliteConnection,
+        new_uuid: &str,
+    ) -> Result<Option<String>, Error> {
+        use schema::rack::dsl;
+        let old_uuid = self.get_rack_uuid(tx)?;
+        diesel::delete(dsl::rack).execute(tx)?;
+        diesel::insert_into(dsl::rack)
+            .values(dsl::uuid.eq(new_uuid))
+            .execute(tx)?;
+        Ok(old_uuid)
+    }
+
+    // Overwrite a share for epoch 0 during rack initialization
+    //
+    // This method should only be called from initialize.
+    fn update_prepare_for_epoch_0(
+        &self,
+        tx: &mut SqliteConnection,
+        share_distribution: SerializableShareDistribution,
+    ) -> Result<(), Error> {
+        use schema::key_shares::dsl;
+        let prepare = KeyShare::new(0, share_distribution)?;
+        diesel::update(dsl::key_shares).set(prepare).execute(tx)?;
+        Ok(())
+    }
+
+    pub fn get_latest_committed_share(
+        &self,
+        conn: &mut SqliteConnection,
+    ) -> Result<(i32, Share), Error> {
+        use schema::key_shares::dsl;
+        let res = dsl::key_shares
+            .select((dsl::epoch, dsl::share))
+            .order(dsl::epoch.desc())
+            .filter(dsl::committed.eq(true))
+            .get_result::<(i32, Share)>(conn)?;
+        Ok(res)
+    }
+
+    pub fn get_committed_share(
+        &self,
+        conn: &mut SqliteConnection,
+        epoch: i32,
+    ) -> Result<Share, Error> {
+        use schema::key_shares::dsl;
+        let share = dsl::key_shares
+            .select(dsl::share)
+            .filter(dsl::epoch.eq(epoch))
+            .filter(dsl::committed.eq(true))
+            .get_result::<Share>(conn)
+            .optional()?;
+
+        match share {
+            Some(share) => Ok(share),
+            None => Err(Error::KeyShareNotCommitted { epoch }),
+        }
+    }
+
+    /// Return true if there is a commit for epoch 0, false otherwise
+    fn is_initialized(&self, tx: &mut SqliteConnection) -> Result<bool, Error> {
+        use schema::key_shares::dsl;
+        Ok(dsl::key_shares
+            .select(dsl::epoch)
+            .filter(dsl::epoch.eq(0))
+            .filter(dsl::committed.eq(true))
+            .get_result::<i32>(tx)
+            .optional()?
+            .is_some())
+    }
+
+    /// Return `Ok(())` if the persisted rack uuid matches `uuid`, or
+    /// `Err(Error::RackUuidMismatch{..})` otherwise.
+    fn validate_rack_uuid(
+        &self,
+        tx: &mut SqliteConnection,
+        rack_uuid: &Uuid,
+    ) -> Result<(), Error> {
+        let rack_uuid = rack_uuid.to_string();
```

Comment:
Do we need to worry about the representation of the UUID changing (e.g., capitalization)? Would it be safer to parse the stored rack UUID into a `Uuid` and compare that? (Should `get_rack_uuid()` return a `Uuid` instead of a `String`?)

Reply:
I don't think there's much to worry about here as a v4 UUID is standardized and the output format default is also standardized and unlikely to change, but it does make me pause for a bit. I did it this way because there is never an error when going from UUID to string, whereas parsing is fallible. This way I don't introduce another error path, although silently failing would be pretty bad! 

I could always implement `ToSql` and `FromSql` impls for a newtype UUID, similar to what I do in the array and bcs macros. What's most unfortunate is that diesel actually [provides these implementations for postgres](https://docs.diesel.rs/2.0.x/diesel/sql_types/struct.Uuid.html), but not sqlite.

Reply:
> I did it this way because there is never an error when going from UUID to string, whereas parsing is fallible.

Normally this would be my biggest concern/annoyance too, but `get_rack_uuid()` already returns a `Result` for other reasons - adding a "someone stored an invalid uuid" error case seems pretty painless.

Reply:
Yeah, you are right. The more I thought about this, the more I agree. I'll add the check.

Reply:
Fixed.